SelectBox: Add boundaries prop to avoid overflow

gregorylegarec opened the pull request:
In Cozy-Harvest-Lib we faced an overflow issue with the SelectBox component:

![capture d ecran 2019-01-16 a 16 47 54](https://user-images.githubusercontent.com/776764/51266826-87899e80-19bc-11e9-942f-d2a52f23bd07.png)


This issue could be resolved in several way:
* Limit the height
* Use a React Portal to show the menu list below all the other element in the DOM

There is already an height limitation in PR #795.

Portaling does not work with preact, and Cozy-Home is still in Preact.

In this PR we propose another solution: give an element as boundaries, and
not expand the menu list outside those boundaries.

With this PR, our first issue is resolved and our SelectBox in Cozy-Harvest-Lib
now looks like this:

![capture d ecran 2019-01-16 a 18 11 55](https://user-images.githubusercontent.com/776764/51266835-8f494300-19bc-11e9-897c-0c2ac67559b1.png)

You can have a preview [here](https://gregorylegarec.github.io/cozy-ui/react/#!/SelectBox/5).

y-lohse commented on react/SelectBox/SelectBox.jsx:
```diff
@@ -204,30 +263,41 @@ class SelectBox extends Component {
+      <div
+        ref={element => {
+          this.element = element
+        }}
```
Why don't we put the ref on the `ReactSelect` directly?

CPatchane replied:
We tried but ReactSelect doesn't forward the ref here, so you got the ReactSelect instance which is not a correct React node

gregorylegarec commented:
I updated this PR with a simpler way to determine space left, based only on `getBoundingClientRect()`, and changed the prop name from `boundaries` to `container`.

@ptbrowne Do you see a way to simplify even more ? How would you test this ?

ptbrowne commented on react/SelectBox/SelectBox.jsx:
```diff
@@ -48,6 +50,44 @@ const customStyles = props => ({
+  const containerPaddingTop =
+    get(actualContainer, 'style.paddingTop') || '0px'
+  const containerPaddingBottom =
+    get(actualContainer, 'style.paddingBottom') || '0px'
```
Why is `get` used ? `style` is always defined on DOM nodes right ?

ptbrowne commented on react/SelectBox/SelectBox.jsx:
```diff
@@ -48,6 +50,44 @@ const customStyles = props => ({
+  if (!(container && selectElement)) return {}
+
+  const isPreactRef = !container.current
+  const actualContainer = isPreactRef ? container : container.current
```
Nit: I personally prefer to shadow `container` in those cases. This makes it clear that we are not going to use `container` afterwards. And it makes it simpler to reason about this function since the number of variable is reduced.

```
container = container.current || container
```

Nit: Since this function's responsibility is about computing heights, I would not put it in charge of knowing the difference between preact and react.

gregorylegarec replied:
It is also bothering me. I will externalize this part, thanks.

ptbrowne commented:
To test this, I would pass mocked nodes to `computedMenuListHeightStyles` and check the result.

gregorylegarec commented:
Updated with externalisation of Preact/React stuff, and some tests.

ptbrowne commented:
LGTM
